File: noita-simulation/classes/solids/dynamic/dynamicSolid.py
from classes.solids.solid import  Solid
from classes.liquids.liquid import Liquid
from classes.solids.static.staticSolid import StaticSolid
from classes.particle import Particle
from random import randint


class DynamicSolid(Solid):
    def __init__(self, i, j) -> None:
        super().__init__(i, j)

    def getNeighbors(self, tiles : list[list]) -> list:
        
        if 0 < self.i < len(tiles)-1 and self.j < len(tiles[0]):
            diagonalLeft = tiles[self.i-1][self.j+1]
            diagonalRight = tiles[self.i+1][self.j+1]
            isDiagonalLeftSolid = isinstance(diagonalLeft, Solid)
            isDiagonalRightSolid = isinstance(diagonalRight, Solid)
        if self.i == 0: #means particle is in the left most column, cant have a diagonalLeft neighbor
            diagonalLeft = None
            isDiagonalLeftSolid = True
            diagonalRight = tiles[self.i+1][self.j+1]
            isDiagonalRightSolid = isinstance(diagonalRight, Solid)
        elif self.i == len(tiles)-1 :#particle at right most column, cant have diagonalRight neighbor
            diagonalRight = None
            isDiagonalRightSolid = True
            diagonalLeft = tiles[self.i-1][self.j+1]
            isDiagonalLeftSolid = isinstance(diagonalLeft, Solid)

        return [diagonalLeft, diagonalRight, isDiagonalLeftSolid, isDiagonalRightSolid]
    
    def update(self, tiles: list[list]) -> list[list]:
        
        try:
            bottom = tiles[self.i][self.j+1]
            
            
            if bottom == 0: 
                self.moveToEmptyCell(self.i, self.j+1, tiles)
                
            
        
            elif isinstance(bottom, Liquid):
                self.switchWithAnotherParticle(self.i, self.j+1, tiles)

            elif isinstance(bottom, Solid):
                
                neighbor = self.getNeighbors(tiles)
                diagonalLeft = neighbor[0]
                diagonalRight = neighbor[1]
                isDiagonalLeftSolid = neighbor[2]
                isDiagonalRightSolid = neighbor[3]

                
    

                if not isDiagonalLeftSolid and not isDiagonalRightSolid:
                     #can go both ways. both diagonalLeft and diagonalRight arent solids
                    direction = randint(0,1)
                    if direction == 0:
                        if isinstance(diagonalLeft, Particle): #if diagonalLeft not empty
                            self.switchWithAnotherParticle(self.i-1, self.j+1, tiles)
                        else:
                            self.moveToEmptyCell(self.i-1, self.j+1, tiles)

                    else:
                        if isinstance(diagonalRight, Particle):
                            self.switchWithAnotherParticle(self.i+1, self.j+1, tiles)
                        else:
                            self.moveToEmptyCell(self.i-1, self.j-1, tiles)


                elif isDiagonalLeftSolid and not isDiagonalRightSolid: #can only move right
                    print(isDiagonalLeftSolid, isDiagonalRightSolid)
                    
                    if isinstance(diagonalRight, Particle): #is either liquid or gas
                        self.switchWithAnotherParticle(self.i+1, self.j+1, tiles)
                    else:
                        self.moveToEmptyCell(self.i+1, self.j+1, tiles)

                elif not isDiagonalLeftSolid and isDiagonalRightSolid: #can only move left
                    
                    if isinstance(diagonalLeft, Particle):
                        self.switchWithAnotherParticle(self.i-1, self.j+1, tiles)
                    else:
                        self.moveToEmptyCell(self.i-1, self.j+1, tiles)
        
        
        except Exception as e:
            pass #print(f"Error: {e}")

            



                
                    



        
    
    

File: noita-simulation-revamped/classes/solids/dynamic/DynamicSolid.java
package classes.solids.dynamic;
import classes.Particle;
import classes.liquids.Liquid;
import classes.solids.Solid;

import java.util.Arrays;
import java.util.Random;

public class DynamicSolid extends Solid{
    int dispersionRate;

    public DynamicSolid(int i, int j){
        super(i, j);
        this.dispersionRate = dispersionRate;
    }






    public Object[][] update(Object[][] tiles){
        try {
            Object bottom = tiles[this.i][this.j+1];
            int bottomI = this.i;
            int bottomJ = this.j+1;

            if (bottom instanceof Integer) {
                tiles = this.moveToEmptyCell(bottomI, bottomJ, tiles);
            }
            else if (bottom instanceof Liquid) {
                tiles = this.switchWithAnotherParticle(bottomI, bottomJ, tiles);
            }
            else if (bottom instanceof Solid) {
                Object[] neighbors = this.getNeighbors(tiles);
                Random r = new Random();
                int direction = r.nextInt(2);
                if (direction == 0){
                    if (this.canGoToCell(this.i-1, this.j+1, tiles)== true){
                        tiles = tryMovingToCell(tiles, this.i-1, this.j+1);
                    } else {
                        if (this.canGoToCell(this.i+1, this.j+1, tiles)== true){
                            tiles = tryMovingToCell(tiles, this.i+1, this.j+1);

                        }
                    }
                } else {
                    if (this.canGoToCell(this.i+1, this.j+1, tiles)== true){
                        tiles = tryMovingToCell(tiles, this.i+1, this.j+1);
                    } else {
                        if (this.canGoToCell(this.i-1, this.j+1, tiles)== true){
                            tiles = tryMovingToCell(tiles, this.i-1, this.j+1);

                        }
                    }
                }
                    

            }
                
            
            
            return tiles;

        }

        catch (Exception e) {
            
            
         }
        

        

        return tiles;
    }
}


            
